bin/cron/pardot_import

```diff
@@ -197,19 +188,189 @@ def main
+  # Mark any email addresses rejected by Pardot as malformed so we don't keep trying to fruitlessly create them forever
+  PEGASUS_DB[:contact_rollups].where("email in ?", malformed_emails).update("email_malformed=true") unless malformed_emails.empty?
+end
+
+# Login to Pardot and request an API key. The API key is valid for (up to) one hour, after which
+# it will become invalid and we will need to request a new one.
+# @returns [String] API key to use for subsequent requests
+def request_pardot_api_key
+  log "Requesting new API key"
+  doc = post_request(PARDOT_AUTHENTICATION_URL,
+    {email: CDO.pardot_username, password: CDO.pardot_password, user_key: CDO.pardot_user_key}
+  )
+
+  status = doc.xpath('/rsp/@stat').text
+  raise "Pardot authentication response failed with status " + status + " " + doc.to_s if status != "ok"
+
+  api_key = doc.xpath('/rsp/api_key').text
+  raise "Pardot authentication response did not include api_key" if api_key.nil?
+
+  $pardot_api_key = api_key
+end
+
+# Build the URL for a batch prospect operation in Pardot. The prospect data becomes a JSON blob in the
+# query param.
+# @param prospects [Array<Hash>] array of hashes of prospect data
+# @param config [Hash] hash of config params to use to control create vs update behavior
+# @returns [String] Pardot URL for API request including encoded JSON prospect data in query string
+def build_batch_prospects_url(prospects, config)
+  prospects_payload_json_encoded = URI.encode({ prospects: prospects }.to_json)
+  # We also need to encode plus signs in email addresses, otherwise Pardot rejects them as invalid.
+  # URI.encode does not encode plus signs, as they are valid characters in the base of a URL.
+  # (Although they are NOT valid in the query string, which is where this data is going.)
+  prospects_payload_json_encoded = prospects_payload_json_encoded.gsub("+", "%2B")
+
+  "#{config[:pardot_url]}?prospects=#{prospects_payload_json_encoded}"
+end
+
+# Make an API request with Pardot authentication, including appending auth params and refreshing Pardot API key and
+# retrying if necessary
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_with_auth_retry(url)
+  # do the post to Pardot
+  post_request_with_auth(url)
+rescue InvalidApiKeyException
+  # If we fail with an invalid key, that probably means our API key (which is good for one hour) has expired. Try one
+  # time to request a new API key and try the post again. If that fails, that is a fatal error.
+  request_pardot_api_key
+  post_request_with_auth(url)
+end
+
+# Make an API request with Pardot authentication
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_request_with_auth(url)
+  request_pardot_api_key if $pardot_api_key.nil?
+  # add the API key and user key parameters to the URL
+  auth_url = append_auth_params_to_url(url)
+  doc = post_request(auth_url, {})
+  doc
```

Comment:
The `doc` variable is unnecessary, simply `post_request(auth_url, {})` accomplishes the same.